src/api.py

```diff
@@ -0,0 +1,220 @@
+        try:
+            return jwt.decode(intoken, SECRET_KEY)
+        except jwt.exceptions.DecodeError:
+            raise Error(FORBIDDEN)
+        except jwt.ExpiredSignatureError:
+            raise Error(UNAUTHORIZED, msg="Signature expired.")
+        except jwt.InvalidTokenError:
+            raise Error(UNAUTHORIZED, msg="Invalid token.")
+
+
+def valid_user(fn):
+    """Decorator to act as middleware, checking token"""
+    def response(self, *args, **kw):
+        if not is_user() or is_admin():
+            raise Error(UNAUTHORIZED, msg="You lack write rights.")
+        return fn(self, *args, **kw)
+    return response
+
+
+def admin_user(fn):
+    """Decorator to act as middleware, checking token for admin rights"""
+    def response(self, *args, **kw):
+        if not is_admin():
+            raise Error(UNAUTHORIZED, msg="You lack admin rights.")
+        return fn(self, *args, **kw)
+    return response
+
+
+def check_token(fn):
+    """Decorator to act as middleware, checking authentication token"""
+    def response(self, *args, **kw):
+        if decode_token(get_token_from_header()):
+            return fn(self, *args, **kw)
+    return response
+
+
+def is_admin():
+    return get_authority_from_token() == 'admin'
+
+
+def is_user():
+    return get_authority_from_token() == 'user'
+
+
+def get_uploader_from_token():
+    return decode_token(get_token_from_header())['sub']
+
+
+def get_authority_from_token():
+    return decode_token(get_token_from_header())['authority']
+
+
+def build_params(filters):
+    if not filters:
+        return "", {}
+    # split by ',' except those preceeded by a top level domain, which will
+    # be a tag URI scheme (e.g. tag:openbookpublishers.com,2009)
+    params    = re.split(r"(?<!\.[a-z]{3}),", filters)
+    options   = {}
+    uris      = []
+    measures  = []
+    countries = []
+    uploaders = []
+    clause  = ""
+    for p in params:
+        try:
+            field, val = p.split(':', 1)
+            if field == "work_uri":
+                uris.append(val)
+            elif field == "measure_uri":
+                measures.append(val)
+            elif field == "country_uri":
+                countries.append(val)
+            elif field == "uploader_uri":
+                uploaders.append(val)
+            else:
+                raise Error(BADFILTERS)
+        except BaseException:
+            raise Error(BADFILTERS, msg="Unknown filter '%s'" % (p))
+
+    process = {"work_uri": uris, "measure_uri": measures,
+               "country_uri": countries, "uploader_uri": uploaders}
+    for key, values in process.items():
+        if len(values) > 0:
+            try:
+                andclause, ops = build_clause(key, values)
+                options.update(ops)
+                clause = clause + andclause
+            except BaseException:
+                raise Error(BADFILTERS)
+
+    return clause, options
+
+
+def build_clause(attribute, values):
+    params = {}
```

Comment:
I can imagine an ORM would make this more readable. I understand what you are trying to achieve, but at a first look, I can't really tell what this is doing.